### ISSUE TITLE
Using PyTorch with Transformers to run inference with 'MPS' backend causes poor results.

### ISSUE THREAD

**Issue by leohpark** [pytorch/pytorch#128435]
### 🐛 Describe the bug

When I run inference using a recent text embedding model, the tensor values and subsequent calculations are not expected values.

Setup:

Python version: 3.12.1 (main, Dec  7 2023, 20:45:44) [Clang 15.0.0 (clang-1500.1.0.2.5)]
PyTorch version: 2.2.2
Transformers version: 4.41.2
NumPy version: 1.26.4
Pandas version: 2.2.1

### code and imports in a Jupyter Notebook in vscode
import torch
import torch.nn.functional as F
from torch import Tensor
from transformers import AutoTokenizer, AutoModel
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

### Set up device for hardware acceleration
device = torch.device('cuda' if torch.cuda.is_available() else 'mps' if torch.backends.mps.is_available() else 'cpu')

### Define the last token pool function
def last_token_pool(last_hidden_states: Tensor, attention_mask: Tensor) -> Tensor:
    left_padding = (attention_mask[:, -1].sum() == attention_mask.shape[0])
    if left_padding:
        return last_hidden_states[:, -1]
    else:
        sequence_lengths = attention_mask.sum(dim=1) - 1
        batch_size = last_hidden_states.shape[0]
        return last_hidden_states[torch.arange(batch_size, device=last_hidden_states.device), sequence_lengths]

### Define the function to create detailed instructions
def get_detailed_instruct(task_description: str, query: str) -> str:
    return f'Instruct: {task_description}\nQuery: {query}'

### Assuming tokenizer and model are already loaded and moved to the device
tokenizer = AutoTokenizer.from_pretrained('Linq-AI-Research/Linq-Embed-Mistral')
model = AutoModel.from_pretrained('Linq-AI-Research/Linq-Embed-Mistral').to(device)
# Move the model to the MPS device
model.to(device)

def nan_test(input, error_text="NaN Nan!"):
    if torch.isnan(input).any():
        print(error_text)
        print(input)

def get_texts():
    task = 'Given a question, retrieve Wikipedia passages that answer the question'
    queries = [
        get_detailed_instruct(task, '최초의 원자력 발전소는 무엇인가?'),
        get_detailed_instruct(task, 'Who invented Hangul?')
    ]
    passages = [
    "현재 사용되는 핵분열 방식을 이용한 전력생산은 1948년 9월 미국 테네시주 오크리지에 설치된 X-10 흑연원자로에서 전구의 불을 밝히는 데 사용되면서 시작되었다. 그리고 1954년 6월에 구소련의 오브닌스크에 건설된 흑연감속 비등경수 압력관형 원자로를 사용한 오브닌스크 원자력 발전소가 시험적으로 전력생산을 시작하였고, 최초의 상업용 원자력 엉더이로를 사용한 영국 셀라필드 원자력 단지에 위치한 콜더 홀(Calder Hall) 원자력 발전소로, 1956년 10월 17일 상업 운전을 시작하였다.",
    "Hangul was personally created and promulgated by the fourth king of the Joseon dynasty, Sejong the Great.[1][2] Sejong's scholarly institute, the Hall of Worthies, is often credited with the work, and at least one of its scholars was heavily involved in its creation, but it appears to have also been a personal project of Sejong."
    ]
    return [*queries, *passages]

def encode_texts(texts, tokenizer, model, device, max_length=4096):
    embeddings = []
    batch_dict = tokenizer(texts, max_length=max_length, padding=True, truncation=True, return_tensors="pt")
    batch_dict = {k: v.to(device) for k, v in batch_dict.items()}
    with torch.no_grad():  # Disable gradient calculation for inference
        outputs = model(**batch_dict)
        embeddings = last_token_pool(outputs.last_hidden_state, batch_dict['attention_mask'])
    return embeddings

def test_vectors(tokenizer=tokenizer, model=model, device=device):
    max_length = 4096
    input_texts = get_texts()
    
    #### Ensure tokenizer handles list of strings
    embeddings = encode_texts(input_texts, tokenizer, model, device, max_length=max_length)

    nan_test(embeddings, "NaN values found in embeddings BEFORE normalization")
    
    embeddings = F.normalize(embeddings, p=2, dim=1)
    nan_test(embeddings, "NaN values found in embeddings AFTER normalization")

    scores = (embeddings[:2] @ embeddings[2:].T) * 100
    nan_test(scores, "NaN values found in scores")

    print(scores.tolist())
    return scores.tolist()

### Run Code

test_scores = test_vectors(tokenizer, model, device)
test_scores

#### [[73.72908782958984, 30.122779846191406], [29.15508460998535, 79.25375366210938]]

Expected Results: Scores using 'mps' backend resemble those from either huggingface example, or cpu.

Actual Result: Scores are not similar.

### Versions

PyTorch version: 2.1.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.1 (x86_64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.8.8 (default, Apr 13 2021, 12:59:45)  [Clang 10.0.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] flake8==3.9.0
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] numpydoc==1.1.0
[pip3] torch==2.1.0
[pip3] torchvision==0.16.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.2.0           hecd8cb5_269  
[conda] mkl-service               2.3.0            py38h9ed2024_1  
[conda] mkl_fft                   1.3.0            py38h4a7008c_2  
[conda] mkl_random                1.2.1            py38hb2f4e1b_2  
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] torch                     2.1.0                    pypi_0    pypi
[conda] torchvision               0.16.0                   pypi_0    pypi

**Comment by skotapati** [comment]
Hello, the issue does not appear to occur with the latest PyTorch nightly, mps returns the expected test_scores. Please update to PyTorch 2.5.0 or later and let us know here if the issue still occurs: https://pytorch.org/get-started/locally/

**Comment by leohpark** [comment]
That sounds awesome, will try out this weekend!

**Comment by skotapati** [comment]
Closing for now, please feel free to reopen if the issue is observed again
